### ISSUE TITLE
SimpleCov showing incorrect coverage result when parallelize is enabled in Rails 6.0.0 beta3

### ISSUE THREAD

**Issue by alkesh26** [colszowka/simplecov#718]
Rails version: 6.0.0 beta3
Ruby version: 2.6.1

**Issue**
We have configured our test helper file as below
```
require "simplecov"

SimpleCov.start do
  add_filter "/test/"

  add_group "Models", "app/models"
end

ENV["RAILS_ENV"] ||= "test"
require_relative "../config/environment"
require "rails/test_help"

class ActiveSupport::TestCase
  parallelize(workers: :number_of_processors)
  
  fixtures :all
end
```

On executing `rake test` the coverage report is incorrect even when test cases are written of the method. So this is our user model
```
class User < ApplicationRecord
  devise :database_authenticatable, :registerable, :trackable,
         :recoverable, :rememberable, :validatable

  has_one_attached :avatar
  validates :email, uniqueness: true

  def admin?
    self.role == "admin"
  end

  def name
    "#{first_name} #{last_name}"
  end
end
```

and the test cases we have are
```
def test_user_admin
    user = users :admin
    assert user.admin?
  end

  def test_user_is_not_an_admin
    user = users :albert
    assert_not user.admin?
  end

  def test_should_return_first_name_and_last_name_as_name
    user = users :albert
    assert_equal "Albert Smith", user.name
  end
```

But the coverage report when `parallelize` is enabled is as below
<img width="602" alt="wrong report" src="https://user-images.githubusercontent.com/1019076/54946619-7fbb1d00-4f5e-11e9-960a-8a6e798b18d2.png">

and when it is commented out the test results are all green
<img width="598" alt="correct report" src="https://user-images.githubusercontent.com/1019076/54946629-88135800-4f5e-11e9-8ada-5c751b4cf48e.png">

I tried below approaches but all of them give correct report only when `parallelize` is commented:
1. https://github.com/colszowka/simplecov/issues/235#issuecomment-22271831
2. Created a `.simplecov` file in root directory and executed `rake`.
3. https://github.com/colszowka/simplecov/issues?utf8=✓&q=is%3Aissue+is%3Aopen+parallel followed the issues and tried other approaches.

But I am still not able to get the correct coverage result when `parallelize` is enabled.

Many thanks in advance.

**Comment by oystersauce8** [comment]
@alkesh26 I don't have a solution for you but have [link](https://github.com/oystersauce8/bug-demo-782735e) to my app and reproducible steps:
  1. Uncomment` parallelize(workers: :number_of_processors)`
  2. Run tests `rm -rf coverage/; RAILS_ENV=test ./bin/rails test; open coverage/index.html`
  3. Coverage of `app/lib/watermelon/example.rb` shows 0%
  4. Comment out `parallelize(workers: :number_of_processors)`
  2. Run tests `rm -rf coverage/; RAILS_ENV=test ./bin/rails test; open coverage/index.html`
  3. Coverage of `app/lib/watermelon/example.rb` shows 70%

**Comment by PragTob** [comment]
Yes this is a biggie. Sorry for taking so long to respond, I took a bit of a break from maintaining simplecov.

Making simplecov completely work with rails and its new parallelization will be a bigger undertaking I fear. PRs are welcome. I'll see if I can allocate the time for it but I'm unsure.

Thank you for reporting :green_heart:

**Comment by bbascarevic-tti** [comment]
As a workaround we can piggy-back on the already existing feature of result merging, we just need to trick SimpleCov that each parallel run is another Command. We set `SimpleCov.use_merging` to `true` in the root process, and then in each fork we change the command name so forks would not overwrite each others' results.

```
# test_helper.rb

require 'simplecov'

SimpleCov.use_merging(true)  # Important!
SimpleCov.start('rails') do
  ... # Whatever you use here...
end

module ActiveSupport
  class TestCase
    parallelize

    parallelize_setup do |worker|
      SimpleCov.command_name("#{SimpleCov.command_name}-#{worker}")
      SimpleCov.pid = Process.pid
      Coverage.start      
    end
    # ...
  end
end
```

**Comment by chrisdebruin** [comment]
@bbascarevic-tti doesn't work for me, Are you running test with `PARALLEL_WORKERS=x` ?
I' am using `parallelize(workers: :number_of_processors)` and it reports wrong coverage.

**Comment by brandoncordell** [comment]
@bbascarevic-tti @chrisdebruin This doesn't work for me either. I'm also using `parallelize(workers: :number_of_processors)`.

**Comment by michaelherold** [comment]
The solution by @bbascarevic-tti is almost there. Here's what I ended up with; it seems to be working as expected:

```ruby
if ENV['COVERAGE']
  require 'simplecov'

  SimpleCov.start 'rails'
end

ENV['RAILS_ENV'] ||= 'test'
require_relative '../config/environment'
require 'rails/test_help'

module ActiveSupport
  class TestCase
    # Run tests in parallel with specified workers
    parallelize(workers: :number_of_processors)

    if ENV['COVERAGE']
      parallelize_setup do |worker|
        SimpleCov.command_name "#{SimpleCov.command_name}-#{worker}"
      end

      parallelize_teardown do |worker|
        SimpleCov.result
      end
    end

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    # Add more helper methods to be used by all tests here...
  end
end
```

There are a few changes from @bbascarevic-tti's solution:

1. Do not set `SimpleCov.pid` to `Process.pid` in the child process. This causes SimpleCov to report a failure for every single child process because SimpleCov then thinks that the child process is the coordinator.
2. Do not specify `SimpleCov.use_merging` since it defaults to `true`.
3. Don't bother wiping out the `SimpleCov.at_exit` hook - these are not run in the workers anyway.
4. Set a `parallelize_teardown` hook to call `SimpleCov.result`. This complex method generates the result and, if `SimpleCov.use_merging` is true (which it is by default), stores the result for later merging. This allows the results of all workers to be merged into a single result.

Here's an example output:

```
↪ rm -rf coverage/ && DISABLE_SPRING=1 COVERAGE=1 bin/rails test
Run options: --seed 52156

# Running:

......................

Finished in 0.393015s, 55.9775 runs/s, 101.7772 assertions/s.
22 runs, 40 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for MiniTest, MiniTest-0, MiniTest-1, MiniTest-2, MiniTest-3 to /home/herold/project/coverage. 58 / 148 LOC (39.19%) covered.
```

I've run the wipe / re-run without Spring many times and receive a consistent result so I don't believe there are any race conditions in the process model. When you mix in Spring, you get the lovely problems inherent in such a venture, so YMMV if you choose to try it with Spring.

## Threads

Note that I've only been able to produce completely consistent results using process parallelization. When I tested out `with: :threads`, I noticed that the coverage numbers would non-deterministically jump between `58 / 148` and `49 / 148`. There were also errors due to thread deadlocks in the database too though, so I think threads on MRI just aren't fully baked (which is fine since you don't get true parallelism anyway).

## Next Steps

I haven't been able to figure out a way to upstream this workflow into SimpleCov, so if someone has any bright ideas, please feel free to do that! The way `parallel_tests` sets the environment variables and gives you access to both the current worker ID and the total number of workers makes sense. Perhaps Rails could use similar functionality to pass the total number of workers into the `parallelize_setup` and `parallelize_teardown` hooks?

**Comment by brandoncordell** [comment]
@michaelherold I think you're close! I'm running into an issue though and maybe it's just general SimpleCov config and not due to parallelize?

Regular tests

    rails test
    Running via Spring preloader in process 77422
    Run options: --seed 42107
    
    # Running:
    
    .........................
    
    Finished in 12.725153s, 1.9646 runs/s, 2.9076 assertions/s.
    25 runs, 37 assertions, 0 failures, 0 errors, 0 skips
    Coverage report generated for MiniTest, MiniTest-0, MiniTest-1 to  /Users/Bcordell/Code/<redacted>/coverage. 177 / 232 LOC (76.29%) covered.

System tests run right afterwards

    rails test:system                                                                                        

    # Running:

    Capybara starting Puma...
    * Version 3.12.1 , codename: Llamas in Pajamas
    * Min threads: 0, max threads: 4
    * Listening on tcp://127.0.0.1:63394
    Capybara starting Puma...
    * Version 3.12.1 , codename: Llamas in Pajamas
    * Min threads: 0, max threads: 4
    * Listening on tcp://127.0.0.1:63392
    ..
    
    Finished in 14.470213s, 0.1382 runs/s, 0.0691 assertions/s.
    2 runs, 1 assertions, 0 failures, 0 errors, 0 skips
    Coverage report generated for MiniTest, MiniTest-0, MiniTest-1 to /Users/Bcordell/Code/<redacted>/coverage. 122 / 293 LOC (41.64%) covered.

Regular tests again run right after the system tests

Regular tests

    rails test
    Running via Spring preloader in process 77422
    Run options: --seed 42107
    
    # Running:
    
    .........................
    
    Finished in 12.725153s, 1.9646 runs/s, 2.9076 assertions/s.
    25 runs, 37 assertions, 0 failures, 0 errors, 0 skips
    Coverage report generated for MiniTest, MiniTest-0, MiniTest-1 to  /Users/Bcordell/Code/<redacted>/coverage. 177 / 232 LOC (76.29%) covered.

It seems like it's not merging my system tests when I run `rails test:system`

**Comment by PragTob** [comment]
Hey everyone,

thanks all of you for your work on this and exchanging ideas helping each other and laying important ground work :clap: 

I don't have a lot of time to invest into this right now (sadly), however I submitted a simplecov project to RubyTogether and this is on the list of things to work on/fix should the project get accepted.

**Comment by PragTob** [comment]
And as a small update, I did get the [ruby together](https://rubytogether.org/) fund and hence an improved rails support especially its new test parallelization is towards the top of the list (after branch coverage and some friends though)

**Comment by AxelTheGerman** [comment]
@PragTob what's the status on this one? Disabling the parallelize feature of Rails is definitely a workaround but would be great to have simplecov work with it - looks like some of the proposals above might be quite close to getting things working properly.

**Comment by PragTob** [comment]
@AxelTheGerman hey, well - in the end as it's always the case with projects it didn't make the cut in the end. There were more branch coverage issues to iron out (and even more now that I finally used it in a project myself). So it's kinda a question of "when I get to it". Truth be told, due to other circumstances I'm not spending too much time on OSS these days. I might bite down some time during a vacation and do it, however, I think I'd first work on polishing existing features (branch coverage...) and bugs. As I'm sure, the parallelize implementation will bring many more fun bugs for me to deal with. Such as #815. Not the answer you were looking for I wager, but sadly the honest one I can give :)

**Comment by AxelTheGerman** [comment]
@PragTob thanks for all the awesome work you do!

I also agree on making existing features robust first - as well as taking care of bugs and tech debt.

I am fine without parallelize support but for now I haven't figured out how to even turn it off properly. If there is a workaround to get rid of parallelization in a Rails 6 app that'd be great. All I want is running regular tests and system tests with combined coverage - which should be possible with `rails test:system test`

I have to check some of my other projects and see what might be different. Will report back when I figure it out :)

**Comment by AxelTheGerman** [comment]
I think the problem was either that I still had `parallelize(workers: 1)` in my test setup OR that I actually didn't have any system tests yet.
As soon as I removed `parallelize` completely and added a working spec, the test coverage is working fine for `rails test:system test` :tada:

**Comment by viral810** [comment]
Omit loading `spring` from `bin/rails` and `bin/rake` from Rails 6 application. Loading spring server has his side effects

```
# bin/rails

#!/usr/bin/env ruby
# load File.expand_path('spring', __dir__)
...
```

**Comment by joshmfrankel** [comment]
This may be common knowledge, but make sure you're also eager_loading your testing environment. I followed the above advice, but still had some accuracy issues until I discovered this. Just run with `CI=true bin/rails test:all` 

```ruby
# config/environments/test.rb
Rails.application.configure do
  # many more configs...
  
  # Use the CI EnvVar to control this
  config.eager_load = ENV["CI"].present?
  
  # some more configs...
end
```

**Comment by searls** [comment]
Any interest in maybe documenting @michaelherold's (excellent! working!) [approach](https://github.com/simplecov-ruby/simplecov/issues/718#issuecomment-538201587) in the README, until/unless built-in support for Rails parallelization is added?

Now that Rails will only start parallelizing tests [once you hit 50 of them](https://world.hey.com/jorge/making-rails-run-just-a-few-tests-faster-2c82dc4b), the fact coverage will suddenly go from correct to near-0% is really confusing and takes time to piece together by searching issues in the repo. (The optimization is totally valid, but when a working initial setup stops working suddenly it can be disorienting.)

**Comment by matteeyah** [comment]
For some reason, I still can't get SimpleCov to report correct percentages - e.g. it's reporting 0% coverage on `ApplicationController`, but properly calculates coverage for most other controllers.

---

This is our `test/test_helper.rb`.

```ruby
# frozen_string_literal: true

if ENV['CI']
  require 'simplecov'
  require 'simplecov-cobertura'
  require 'minitest/reporters'

  SimpleCov.start 'rails' do
    command_name = ENV.fetch('CI_JOB_NAME_SLUG', nil) # rubocop:disable Lint/UselessAssignment

    add_filter '/app/admin'
    add_filter 'vendor'

    add_group 'Policies', 'app/policies'

    formatter SimpleCov::Formatter::CoberturaFormatter
  end

  Minitest::Reporters.use!([
                             Minitest::Reporters::DefaultReporter.new,
                             Minitest::Reporters::JUnitReporter.new('test/reports', true, single_file: true)
                           ])
end

ENV['RAILS_ENV'] ||= 'test'
require_relative '../config/environment'
require 'rails/test_help'

require 'webmock/minitest'
require 'sidekiq/testing'

module ActiveSupport
  class TestCase
    # Run tests in parallel with specified workers
    parallelize(workers: :number_of_processors)

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    if ENV['CI']
      parallelize_setup { |worker| SimpleCov.command_name "#{SimpleCov.command_name}-#{worker}" }
      parallelize_teardown { |_| SimpleCov.result }
    end

    teardown do
      Redis.new(url: ENV.fetch('REDIS_URL', 'redis://localhost:6379/1')).flushdb
    end

    # Add more helper methods to be used by all tests here...
  end
end

# Disable external requests
WebMock.disable_net_connect!(
  allow_localhost: true,
  allow: ['https://googlechromelabs.github.io', 'https://edgedl.me.gvt1.com']
)

# Use OmniAuth in test mode
OmniAuth.config.test_mode = true

Sidekiq::Testing.fake!
```

This is how we collate the results

```ruby
namespace :coverage do
  task report: :environment do
    require 'simplecov'

    SimpleCov.collate Dir['coverage/.resultset*.json'], 'rails' do
      formatter SimpleCov::Formatter::HTMLFormatter
    end
  end
end
```

**Comment by sferik** [comment]
Thanks for the detailed report and to everyone in the thread who built up the manual `parallelize_setup`/`parallelize_teardown workaround`, which directly informed the eventual built-in support.

SimpleCov now handles `ActiveSupport::TestCase.parallelize(workers: ...)` forks automatically:

* 94eca165ce5456c3631565c1fa4a6b9a63e70c06 added `SimpleCov.enable_for_subprocesses` and `SimpleCov.at_fork`, with a default `at_fork` block that restarts SimpleCov in the child with a unique command_name so per-worker resultsets merge cleanly.
* 2508e4eb0f7694dc66e62f36d23219befecfde5d moved the hook to `Process._fork`, which is the canonical extension point in Ruby 3.1 and it composes correctly with other gems hooking the same point.
* cf08ec13048c0f9b9f64e72dafb6beb76c311475 turned subprocess tracking on by default in the bundled rails profile.

So for your original `test_helper`, switching `SimpleCov.start do ... end` to `SimpleCov.start 'rails' do ... end` is enough. The parallelize forks will contribute to the merged report, no `parallelize_setup`/`parallelize_teardown` boilerplate is needed. If you'd rather keep a bare `SimpleCov.start`, adding `enable_for_subprocesses true` inside the block has the same effect.

I’m going to close this issue now but please don’t hesitate to reopen it (or open a fresh issue) if you (or anyone reading this) still see workers dropping coverage. Thanks again for the report!

**Comment by MatheusRich** [comment]
Lovely, @sferik! is this released already?

**Comment by tute** [comment]
Thank you! Bundling from GitHub (SHA 1f0847c00bff611062f603dd11d62c9c4738b9f8) on a close-to-greenfield app I have, I need to generate the report at least once with `parallelize_teardown { SimpleCov.result }`. Is it expected?

**Comment by sferik** [comment]
> Lovely, @sferik! is this released already?

I just merged the code yesterday, so it’s not yet released, but you can test it out by putting the following in your `Gemfile`:

```ruby
gem "simplecov", git: "https://github.com/simplecov-ruby/simplecov.git", branch: "main", require: false
```

If you decide to do so, please confirm that the bug is fixed for you.

**Comment by sferik** [comment]
> Thank you! Bundling from GitHub (SHA [1f0847c](https://github.com/simplecov-ruby/simplecov/commit/1f0847c00bff611062f603dd11d62c9c4738b9f8)) on a close-to-greenfield app I have, I need to generate the report at least once with `parallelize_teardown { SimpleCov.result }`. Is it expected?

@tute thanks so much for testing out pre-release code! On current SimpleCov with `Process._fork`, it shouldn’t be needed. `parallelize_teardown { SimpleCov.result }` was the workaround before `Process._fork`. The hook should now do that work on its own.

A couple of things worth ruling out on your end:

1. Test count vs. parallelize threshold. `ActiveSupport::TestCase.parallelize` defaults to `threshold: 50`. On a close-to-greenfield app with under 50 tests, Rails silently skips forking and runs everything in the parent. That would look like “I needed `parallelize_teardown` to get anything,” when really the parent’s own `at_exit` is doing the work.
2. `SimpleCov.start "rails"` flips on `enable_for_subprocesses` but a bare `SimpleCov.start` does not, so the fork hook never installs and workers’ coverage genuinely is dropped.

If you can post your `test_helper.rb` and how many tests you have, I can dig in further.

Thanks again for testing this!

**Comment by tute** [comment]
Thank you for your work here (and probably elsewhere too :-) ).

This snippet shows hundreds of tests being split in several subprocesses. The first run either has the line I quote above, or was run with `coverage/` files already built; the second one is run without that line and without `coverage/` files (reruns don't improve it):
```
tc-simplecov % rake test:all
Rebuilding...

Done in 514ms.
Running 1109 tests in parallel using 16 processes
Run options: --seed 34670

# Running:

.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 3.008990s, 368.5623 runs/s, 1154.5404 assertions/s.
1109 runs, 3474 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for Minitest, Minitest (subprocess: 32237), Minitest (subprocess: 32240), Minitest (subprocess: 32244), Minitest (subprocess: 32248), Minitest (subprocess: 32252), Minitest (subprocess: 32256), Minitest (subprocess: 32260), Minitest (subprocess: 32264), Minitest (subprocess: 32268), Minitest (subprocess: 32272), Minitest (subprocess: 32276), Minitest (subprocess: 32280), Minitest (subprocess: 32284), Minitest (subprocess: 32288), Minitest (subprocess: 32292), Minitest (subprocess: 32296), Minitest (subprocess: 32457), Minitest (subprocess: 32460), Minitest (subprocess: 32465), Minitest (subprocess: 32468), Minitest (subprocess: 32472), Minitest (subprocess: 32476), Minitest (subprocess: 32480), Minitest (subprocess: 32485), Minitest (subprocess: 32488), Minitest (subprocess: 32492), Minitest (subprocess: 32496), Minitest (subprocess: 32501), Minitest (subprocess: 32504), Minitest (subprocess: 32508), Minitest (subprocess: 32512), Minitest (subprocess: 32516) to ./coverage. 4135 / 4410 LOC (93.76%) covered.

tc-simplecov % rm -rf coverage
tc-simplecov % rake test:all

Rebuilding...

Done in 527ms.
Running 1109 tests in parallel using 16 processes
Run options: --seed 9833

# Running:

.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

Finished in 2.801397s, 395.8740 runs/s, 1240.0958 assertions/s.
1109 runs, 3474 assertions, 0 failures, 0 errors, 0 skips
Coverage report generated for Minitest to ./coverage. 1489 / 4371 LOC (34.06%) covered.
Line coverage (34.06%) is below the expected minimum coverage (90.00%).
  Lowest-coverage files (line):
      ...
SimpleCov failed with exit 2 due to a coverage related error
```
Here's my `test_helper.rb` contents:
```ruby
tc-simplecov % cat test/test_helper.rb
ENV["RAILS_ENV"] ||= "test"

require "simplecov"
SimpleCov.start(:rails) do
  minimum_coverage 90
  enable_for_subprocesses true
end
require_relative "../config/environment"

require "rails/test_help"
require "mocha/minitest"
require "webmock/minitest"
require_relative "support/env_helpers"
require_relative "support/vcr_setup"

Rails.application.eager_load!

FHIR.logger.level = Logger::WARN

# Prevent database truncation if the environment is production
abort("The Rails environment is running in production mode!") if Rails.env.production?

module ActiveSupport
  class TestCase
    # parallelize_teardown { SimpleCov.result } # THIS IS THE LINE I QUOTE AND UNQUOTE FOR TESTING THIS ISSUE
    parallelize(workers: :number_of_processors)

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    # Reset Rack::Attack cache before each test
    setup do
      Rack::Attack.reset!
    end

    # Add more helper methods to be used by all tests here...
    include EnvHelpers
  end
end

module ActionDispatch
  class IntegrationTest
    # Reset Rack::Attack cache before each integration test
    setup do
      Rack::Attack.reset!
    end
  end
end
```
